pennylane/templates/subroutines/qsvt.py

```diff
@@ -0,0 +1,279 @@
+    ...     qml.qsvt(A, angles, wires=[0, 1])
+    ...     return qml.expval(qml.PauliZ(wires=0))
+    >>> print(qml.draw(example_circuit)(A))
+    0: ─╭QSVT─┤  <Z>
+    1: ─╰QSVT─┤
+    >>> print(qml.draw(example_circuit, expansion_strategy="device")(A))
+    0: ─╭∏_ϕ(0.10)─╭BlockEncode(M0)─╭∏_ϕ(0.20)─╭BlockEncode(M0)†─╭∏_ϕ(0.30)─┤  <Z>
+    1: ─╰∏_ϕ(0.10)─╰BlockEncode(M0)─╰∏_ϕ(0.20)─╰BlockEncode(M0)†─╰∏_ϕ(0.30)─┤
+    """
+    if qml.math.shape(A) == () or qml.math.shape(A) == (1,):
+        A = qml.math.reshape(A, [1, 1])
+
+    c, r = qml.math.shape(A)
+
+    UA = BlockEncode(A, wires=wires, do_queue=False)
+    projectors = []
+
+    if convention == "Wx":
+        angles = _qsp_to_qsvt(angles)
+        global_phase = (len(angles) - 1) % 4
+
+        if global_phase:
+            qml.exp(qml.Identity(wires=wires), 0.5j * np.pi * (4 - global_phase))
+
+    for idx, phi in enumerate(angles):
+        dim = c if idx % 2 else r
+        projectors.append(PCPhase(phi, dim=dim, wires=wires, do_queue=False))
+
+    return QSVT(UA, projectors, wires=wires)
+
+
+class QSVT(Operation):
+    r"""Implements the
+    `quantum singular value transformation <https://arxiv.org/abs/1806.01838>`__ circuit.
+
+    Given an :class:`~.Operator` :math:`U`, which block encodes the matrix :math:`A`, and a list of
+    projector-controlled phase shift operations :math:`\vec{\Pi}_\phi`, this template applies a
+    circuit for the quantum singular value transformation as follows.
+
+    When the number of projector-controlled phase shifts is even (:math:`d` is odd), the QSVT
+    circuit is defined as:
+
+    .. math::
+
+        U_{QSVT} = \Pi_{\phi_1}U\left[\prod^{(d-1)/2}_{k=1}\Pi_{\phi_{2k}}U^\dagger
+        \Pi_{\phi_{2k+1}}U\right]\Pi_{\phi_{d+1}}.
+
+
+    And when the number of projector-controlled phase shifts is odd (:math:`d` is even):
+
+    .. math::
+
+        U_{QSVT} = \left[\prod^{d/2}_{k=1}\Pi_{\phi_{2k-1}}U^\dagger\Pi_{\phi_{2k}}U\right]
+        \Pi_{\phi_{d+1}}.
+
+    This circuit applies a polynomial transformation (:math:`Poly^{SV}`) to the singular values of
+    the block encoded matrix:
+
+    .. math::
+
+        \begin{align}
+             U_{QSVT}(A, \vec{\phi}) &=
+             \begin{bmatrix}
+                Poly^{SV}(A) & \cdot \\
+                \cdot & \cdot
+            \end{bmatrix}.
+        \end{align}
+
+    .. seealso::
+
+        :func:`~.qsvt` and `A Grand Unification of Quantum Algorithms <https://arxiv.org/pdf/2105.02859.pdf>`_
+
+    Args:
+        UA (Operator): the block encoding circuit, specified as an :class:`~.Operator`,
+            like :func:`~.BlockEncode`
+        projectors (Sequence[Operator]): a list of projector-controlled phase
+            shifts that implement the desired polynomial
+        wires (Iterable): the wires the template acts on
+
+    Raises:
+        ValueError: if the input block encoding is not an operator
+
+    **Example**
+
+    >>> dev = qml.device("default.qubit", wires=2)
```

Comment:
This applies to all example code: I think they look better with more vertical space in between sections